Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -1345,14 +1384,25 @@ extension HTTP2FrameDecoder {
     ///
     /// Sadly, because it's generic and has a closure, we need to force it to be inlined at all call sites, which is
     /// not ideal.
+    ///
+    /// The `body` function must assign a value to the parser state, otherwise it will remain the intermediary
+    /// `appending` state, which is not wanted and it would result in the failure of the final assertion.
+    /// If the state doesn't get updated because of an error that was thrown, this function sets the parser state
+    /// to the initial one.
     @inline(__always)
-    private mutating func avoidingParserCoW<ReturnType>(_ body: (inout ParserState) -> ReturnType) -> ReturnType {
+    private mutating func avoidingParserCoW<ReturnType>(_ body: (inout ParserState) throws -> ReturnType) rethrows -> ReturnType {
+        let initialState = self.state
```

Comment:
Sorry I didn't catch this before, but I think by doing this we're undoing the effectiveness of this function. The reason is that the closure will be capturing the associated payload held by `self.state`, now we're grabbing it here again so we'll CoW if we mutate the one we captured in the closure.

I don't think restoring the state in case of an error is actually necessary either: the previous code didn't do it.

Reply:
You are right, I'm trying to think how to solve this. The problem is that in the case of an error that assertion that checks the final state fails - the previous implementation didn't receive a throwing function as a parameter so the state was always updated to a valid one (in the body of that function). In the `append()` function where `avoidingParserCoW` was already used there are no throwing function calls.

Reply:
should we get rid of that assertion if the function is throwing?

Reply:
> should we get rid of that assertion if an error is thrown?

No the assertion is there for a reason: we must never leave state in the `appending` state, otherwise the next call which switches over the state won't have any idea what it can do!

Reply:
Thanks!